backend/chainlit/data/sql_alchemy.py

```diff
@@ -0,0 +1,496 @@
+            raise ValueError("No authenticated user in context")
+        if isinstance(element, Avatar): # Skip creating elements of type avatar
+            return
+        if not self.storage_provider:
+            logger.warn(f"SQLAlchemy: create_element error. No blob_storage_client is configured!")
+            return
+        if not element.for_id:
+            return
+
+        content: Optional[Union[bytes, str]] = None
+
+        if element.path:
+            async with aiofiles.open(element.path, "rb") as f:
+                content = await f.read()
+        elif element.url:
+            async with aiohttp.ClientSession() as session:
+                async with session.get(element.url) as response:
+                    if response.status == 200:
+                        content = await response.read()
+                    else:
+                        content = None
+        elif element.content:
+            content = element.content
+        else:
+            raise ValueError("Element url, path or content must be provided")
+        if content is None:
+            raise ValueError("Content is None, cannot upload file")
+
+        context_user = context.session.user
+
+        user_folder = getattr(context_user, 'id', 'unknown')
+        file_object_key = f"{user_folder}/{element.id}" + (f"/{element.name}" if element.name else "")
+
+        if not element.mime:
+            element.mime = "application/octet-stream"
+
+        uploaded_file = await self.storage_provider.upload_file(object_key=file_object_key, data=content, mime=element.mime, overwrite=True)
+        if not uploaded_file:
+            raise ValueError("SQLAlchemy Error: create_element, Failed to persist data in storage_provider")
+
+        element_dict: ElementDict = element.to_dict()
+
+        element_dict['url'] = uploaded_file.get('url')
+        element_dict['objectKey'] = uploaded_file.get('object_key')
+        element_dict_cleaned = {k: v for k, v in element_dict.items() if v is not None}
+
+        columns = ', '.join(f'"{column}"' for column in element_dict_cleaned.keys())
+        placeholders = ', '.join(f':{column}' for column in element_dict_cleaned.keys())
+        query = f"INSERT INTO elements ({columns}) VALUES ({placeholders})"
+        await self.execute_sql(query=query, parameters=element_dict_cleaned)
+
+    @queue_until_user_message()
+    async def delete_element(self, element_id: str):
+        logger.info(f"SQLAlchemy: delete_element, element_id={element_id}")
+        query = """DELETE FROM elements WHERE "id" = :id"""
+        parameters = {"id": element_id}
+        await self.execute_sql(query=query, parameters=parameters)
+
+    async def delete_user_session(self, id: str) -> bool:
+        return False # Not sure why documentation wants this
+
+    async def get_all_user_threads(self, user_id: Optional[str] = None, thread_id: Optional[str] = None) -> Optional[List[ThreadDict]]:
```

Comment:
I'm not sure how chainlit is using this function, but if they are using this for displaying the list of threads only, then I think we should not query all the information inside each thread. That should happen when user click to resume the chat.

Reply:
This function expects a List[ThreadDict] which is basically everything to your point. I was able to speed it up a lot by querying it all at once per user instead of recursively calling each thread, steps in that thread, etc to build it.

This is also used for the search and feedback.

I added a user_thread_limit to cap threads since it'll only grow. I have several users filling up my limit (100 threads) and performance is fine. Are you seeing performance problems or just asking the Chainlit devs why it's like this?

Reply:
This function isn't part of `BaseDataLayer` so it can be defined and used from within this custom data layer as needed.

@sandangel I agree that there might be some performance improvements to be done, I think it's fine for the first release of this custom data layer 👍